src/access/examples/ExampleAccessControlDelegate.cpp

```diff
@@ -0,0 +1,1072 @@
+        {
+            return mStorage->mTargets[index].Get(target);
+        }
+        return CHIP_ERROR_SENTINEL;
+    }
+
+    CHIP_ERROR SetTarget(size_t index, const Target & target) override
+    {
+        ReturnErrorCodeIf(EnsureInPool(), CHIP_ERROR_BUFFER_TOO_SMALL);
+        if (index < EntryStorage::kMaxTargets)
+        {
+            return mStorage->mTargets[index].Set(target);
+        }
+        return CHIP_ERROR_SENTINEL;
+    }
+
+    CHIP_ERROR AddTarget(size_t * index, const Target & target) override
+    {
+        ReturnErrorCodeIf(EnsureInPool(), CHIP_ERROR_BUFFER_TOO_SMALL);
+        size_t count;
+        GetTargetCount(count);
+        if (count < EntryStorage::kMaxTargets)
+        {
+            CHIP_ERROR err = mStorage->mTargets[count].Add(target);
+            if (err == CHIP_NO_ERROR && index != nullptr)
+            {
+                *index = count;
+            }
+            return err;
+        }
+        return CHIP_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    CHIP_ERROR RemoveTarget(size_t index) override
+    {
+        ReturnErrorCodeIf(EnsureInPool(), CHIP_ERROR_BUFFER_TOO_SMALL);
+        size_t count;
+        GetTargetCount(count);
+        if (index < count)
+        {
+            const auto & dest = mStorage->mTargets + index;
+            const auto & src  = dest + 1;
+            auto bytes        = (count - index - 1) * sizeof(*dest);
+            memmove(dest, src, bytes);
+            if (count == EntryStorage::kMaxTargets)
+            {
+                mStorage->mTargets[EntryStorage::kMaxTargets - 1].Clear();
+            }
+        }
+        return CHIP_ERROR_SENTINEL;
+    }
+
+public:
+    void Init(Entry & entry, EntryStorage & storage)
+    {
+        entry.SetDelegate(*this);
+        storage.Init();
+        mEntry   = &entry;
+        mStorage = &storage;
+    }
+
+    bool InUse() const { return mStorage != nullptr; }
+
+    const EntryStorage * GetStorage() const { return mStorage; }
+
+    EntryStorage * GetStorage() { return mStorage; }
+
+    void FixAfterDelete(EntryStorage & storage)
+    {
+        constexpr auto & acl = EntryStorage::acl;
+        constexpr auto end   = acl + ArraySize(acl);
+        if (mStorage == &storage)
+        {
+            mEntry->ResetDelegate();
+        }
+        else if (&storage < mStorage && mStorage < end)
+        {
+            mStorage--;
+        }
+    }
+
+    bool EnsureInPool()
+    {
+        if (mStorage->InPool())
+        {
+            return false;
+        }
+        else if (auto storage = EntryStorage::Find(nullptr))
+        {
+            *storage = *mStorage;
+            mStorage = storage;
+            return false;
+        }
+        return true;
+    }
+
+private:
+    Entry * mEntry          = nullptr;
+    EntryStorage * mStorage = nullptr;
+};
+
+class EntryIteratorDelegate : public EntryIterator::Delegate
+{
+public:
+    // Pool support
+    static EntryIteratorDelegate pool[kEntryIteratorDelegatePoolSize];
```

Comment:
I do wonder: should we use one of our actual pool classes?

Reply:
That is a good question.  Looking into this is followup fodder anyway; no changes needed in this PR.